src/KubernetesClient/Watcher.cs

```diff
@@ -174,6 +150,70 @@ private async Task WatcherLoop(CancellationToken cancellationToken)
             }
         }
 
+        internal static async IAsyncEnumerable<(WatchEventType, T)> CreateWatchEventEnumerator(
+            Func<Task<TextReader>> streamReaderCreator,
+            Action<Exception> onError = null,
+            [EnumeratorCancellation] CancellationToken cancellationToken = default)
+        {
+            Task<TR> AttachCancellationToken<TR>(Task<TR> task)
+            {
+                if (!task.IsCompleted)
+                {
+                    // here to pass cancellationToken into task
+                    return task.ContinueWith(t => t.GetAwaiter().GetResult(), cancellationToken);
+                }
+
+                return task;
+            }
+
+            using var streamReader = await AttachCancellationToken(streamReaderCreator()).ConfigureAwait(false);
```

Comment:
This looks a bit convoluted. It's because StreamReader doesn't support cancellation token right?

Reply:
disposing watcher will dispose the stream which would cancel the `readlineasync` in the original watcher pattern
both `streamReaderCreator()` and `readlineasync` does not honor the cancellation token, here is a workaround to attach it

Reply:
Interesting, haven't seen this ContinueWith/CancellationToken pattern before. I'll look into that a little closer before commenting.

Reply:
Looking into this it seems the underlying task remains active even though the wrapping task becomes cancelled. I think you'd be okay without this pattern.

In a perfect world ReadLineAsync would take a CT... Short of that, to cover the `ReadLineAsync` it might be better to register the CT as a call to stream.Dispose. That will have the same effect as before - ensuring the underlying call waiting on network unwinds immediately.

```csharp
using var streamReader = await AttachCancellationToken(streamReaderCreator()).ConfigureAwait(false);
using var cancellationTokenRegistration = cancellationToken.Register(streamReader.Dispose);
```

For the first call down to content `ReadAsStreamAsync` --- that one doesn't take a CT either. Good news is it doesn't seem like that code path enters a wait state? In the original code it seems like that wasn't covered by cancellation either. So should be okay to leave the `streamReaderCreator()` unguarded by CT.

Reply:
yup 
`streamReaderCreator` is under our control and easy to make it take CT

I am thinking that `AttachCancellationToken` is a waste if I use it only once :)